mujoco_py/mjpid.pyx

```diff
@@ -1,21 +1,78 @@
 from mujoco_py.generated import const
 
-"""
-  Kp == Kp
-  Ki == Kp/Ti
-  Kd == Kp*Td
+cdef int CONTROLLER_TYPE_PI_CASCADE = 1,
 
-  In this situation, Kp is a knob to tune the agressiveness, wheras Ti and Td will
-  change the response time of the system in a predictable way. Lower Ti or Td means
-  that the system will respond to error more quickly/agressively.
+cdef struct PIDErrors:
+    double error
+    double integral_error
+    double derivative_error
 
-  error deadband: if set will shrink error within to 0.0
-  clamp on integral term:  helps on saturation problem in I.
-  derivative smoothing term:  reduces high frequency noise in D.
+cdef struct PIDOutput:
+    PIDErrors errors
+    double output
+
+cdef struct PIDParameters:
+    double dt_seconds  # PID sampling time.
+    double setpoint
+    double feedback
+    double Kp
+    double error_deadband
+    double integral_max_clamp
+    double integral_time_const
+    double derivative_gain_smoothing
+    double derivative_time_const
+    PIDErrors previous_errors
+
+cdef mjtNum c_zero_gains(const mjModel*m, const mjData*d, int id) with gil:
+    return 0.0
+
+cdef PIDOutput _pid(PIDParameters parameters):
+    """
+    A general purpose PID controller implemented in the standard form. 
+    Kp == Kp
+    Ki == Kp/Ti
+    Kd == Kp*Td
+    
+    In this situation, Kp is a knob to tune the agressiveness, wheras Ti and Td will
+    change the response time of the system in a predictable way. Lower Ti or Td means
+    that the system will respond to error more quickly/agressively.
+    
+    error deadband: if set will shrink error within to 0.0
+    clamp on integral term:  helps on saturation problem in I.
+    derivative smoothing term:  reduces high frequency noise in D.
+    
+    :param parameters: PID parameters
+    :return: A PID output struct containing the control output and the error state
+    """
+
+    cdef double error = parameters.setpoint - parameters.feedback
+
+    # clamp error that's within the error deadband
+    if fabs(error) < parameters.error_deadband:
+        error = 0.0
+
+    # compute derivative error
+    cdef double derivative_error = (error - parameters.previous_errors.error) / parameters.dt_seconds
+
+    derivative_error = (1.0 - parameters.derivative_gain_smoothing) * parameters.previous_errors.derivative_error + \
+                       parameters.derivative_gain_smoothing * derivative_error
+
+    cdef double derivative_error_term = derivative_error * parameters.derivative_time_const
+
+    # update and clamp integral error
+    integral_error = parameters.previous_errors.integral_error
+    integral_error += error * parameters.dt_seconds
+    integral_error = fmax(-parameters.integral_max_clamp, fmin(parameters.integral_max_clamp, integral_error))
+
+    cdef double integral_error_term = 0.0
+    if parameters.integral_time_const != 0:
+        integral_error_term = integral_error / parameters.integral_time_const
```

Comment:
Are you sure you don't want to clamp after `integral_error / parameters.integral_time_const` ?

Reply:
I'm not changing the implementation in this PR. Maybe @bayesian has context?

Reply:
integral_time_const is a constant, clamp before or after multiplied by a constant is equivalent, it only affects the clamp range. It is likely this logic is from our c++ controller.

Reply:
Except that one is easier to think about than the other.